src/main/java/net/minecraftforge/client/gui/config/ControlCreator.java

```diff
@@ -0,0 +1,525 @@
+            boolean oldValue = state.getValue();
+            boolean newValue = !oldValue;
+            state.setValue(newValue);
+            boolean isValid = interactor.isValid(newValue);
+            if (isValid)
+                interactor.save(newValue);
+            interactor.onUpdate(isValid, newValue);
+        });
+
+        interactor.addUpdateResponder((isValid, newValue) -> {
+            state.setValue(newValue);
+            Widget control = interactor.control;
+            control.func_238482_a_(new StringTextComponent(Boolean.toString(newValue)));
+            if (!isValid)
+                control.setFGColor(RED);
+            else
+                control.setFGColor(newValue ? GREEN : RED);
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T extends Number> void createNumericInteractionWidget(Interactor<T> interactor) {
+        T value = interactor.initialValue;
+        Range<? extends Number> range = getRangeForSliderCreation(interactor);
+        if (range != null) {
+            if (value instanceof Integer)
+                createSliderNumericInteractionWidget((Interactor<Integer>) interactor, range, d -> Math.round(Math.round(d)));
+            else if (value instanceof Long)
+                createSliderNumericInteractionWidget((Interactor<Long>) interactor, range, Math::round);
+            else if (value instanceof Double)
+                createSliderNumericInteractionWidget((Interactor<Double>) interactor, range, Function.identity());
+            else
+                throw new IllegalStateException("Don't know how to make a widget for Ranged Number " + value);
+        } else {
+            if (value instanceof Integer)
+                createTextualNumericInteractionWidget((Interactor<Integer>) interactor, Integer::parseInt, getLongestValueStringLength(Integer.MIN_VALUE));
+            else if (value instanceof Long)
+                createTextualNumericInteractionWidget((Interactor<Long>) interactor, Long::parseLong, getLongestValueStringLength(Long.MIN_VALUE));
+            else if (value instanceof Double)
+                createTextualNumericInteractionWidget((Interactor<Double>) interactor, Double::parseDouble, getLongestValueStringLength(-Double.MAX_VALUE));
+            else
+                throw new IllegalStateException("Don't know how to make a widget for Non-ranged Number " + value);
+        }
+    }
+
+    /**
+     * @return null if a slider should not be created for the interactor.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T extends Number, C extends Number & Comparable<? super C>> Range<C> getRangeForSliderCreation(Interactor<T> interactor) {
+        Range<C> range = (Range<C>) interactor.getData(RANGE_KEY);
+        if (range == null)
+            return null;
+        double min = range.getMin().doubleValue();
+        double max = range.getMax().doubleValue();
+        if (max - min > 10)
```

Comment:
This seems like an arbitary bound... At least put it into a `private static final int` and add some clarification why 10 seems like a good value

Reply:
It’s very arbitrary, I’m not unhappy about it though - if people don’t like it I want them to change it by supplying their own ControlCreator for their config screen. The aim then is to make it as easy as possible to change which I think I have done.